azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/AzureComputeServiceAdapter.java

```diff
@@ -0,0 +1,429 @@
+import org.jclouds.domain.LoginCredentials;
+import org.jclouds.logging.Logger;
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import org.jclouds.util.Predicates2;
+
+/**
+ * Defines the connection between the {@link AzureComputeApi} implementation and the jclouds
+ * {@link org.jclouds.compute.ComputeService}.
+ */
+@Singleton
+public class AzureComputeServiceAdapter implements ComputeServiceAdapter<VMDeployment, VMHardware, VMImage, Location> {
+
+   private static int runningNumber = 1;
+   private final String azureGroup;
+
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   private Logger logger = Logger.NULL;
+
+   private final AzureComputeApi api;
+
+   private final AzureComputeConstants azureComputeConstants;
+
+   @Inject
+   AzureComputeServiceAdapter(final AzureComputeApi api, final AzureComputeConstants azureComputeConstants,
+                              @Named("azureGroupId") String group) {
+
+      this.api = api;
+      this.azureComputeConstants = azureComputeConstants;
+      this.azureGroup = group;
+   }
+
+   @Override
+   public NodeAndInitialCredentials<VMDeployment> createNodeWithGroupEncodedIntoName(
+           final String group, final String name, final Template template) {
+
+      final String location = template.getLocation().getId();
+      DeploymentTemplateBuilder deploymentTemplateBuilder = api.deploymentTemplateFactory().create(group, name, template);
+
+      final String loginUser = deploymentTemplateBuilder.getLoginUserUsername();
+      final String loginPassword = deploymentTemplateBuilder.getLoginPassword();
+
+      DeploymentBody deploymentTemplateBody =  deploymentTemplateBuilder.getDeploymentTemplate();
+
+      DeploymentProperties properties = DeploymentProperties.create(deploymentTemplateBody);
+
+      final String deploymentTemplate = UrlEscapers.urlFormParameterEscaper().escape(deploymentTemplateBuilder.getDeploymentTemplateJson(properties));
+
+      logger.debug("Deployment created with name: %s", name);
+
+      final Set<VMDeployment> deployments = Sets.newHashSet();
+
+      ResourceGroupApi resourceGroupApi = api.getResourceGroupApi();
+      ResourceGroup resourceGroup = resourceGroupApi.get(azureGroup);
+      final String resourceGroupName;
+
+      if (resourceGroup == null){
+         final Map<String, String> tags = ImmutableMap.of("description", "jClouds managed VMs");
+         resourceGroupName = resourceGroupApi.create(azureGroup, location, tags).name();
+      } else {
+         resourceGroupName = resourceGroup.name();
+      }
+
+      if (!retry(new Predicate<String>() {
+         @Override
+         public boolean apply(final String name) {
+            runningNumber++;
+
+            Deployment deployment = api.getDeploymentApi(resourceGroupName).create(name, deploymentTemplate);
```

Comment:
Get the deployment api outside the loop to save the class lookup in every iteration.